_posts/2022-02-24-torchserve-grpc.md

```diff
@@ -0,0 +1,217 @@
+
+
+<div class="table-wrapper" markdown="block">
+
+|      | B64 (0.306 MB) | Binary (0.229) |
+|------|----------------|----------------|
+| REST | 0.884          | 0.628          |
+| gRPC |       X        | 0.645          |
+
+<p>Table 4. Results for small images: 360x640</p>
+
+</div>
+
+<div class="table-wrapper" markdown="block">
+
+|      | B64 (0.306 MB) | Binary (0.229) |
+|------|----------------|----------------|
+| REST | 1.262          | 0.946          |
+| gRPC |        X       | 0.927          |
+
+<p>Table 5. Results for medium images: 720x1280</p>
+
+</div>
+
+<div class="table-wrapper" markdown="block">
+
+|      | B64 (0.306 MB) | Binary (0.229) |
+|------|----------------|----------------|
+| REST | 2.188          | 1.384          |
+| gRPC |         X      | 1.422          |
+
+<p>Table 6. Results for large images: 1080x1920</p>
+
+</div>
+
+Translating the insights gained from benchmarking with the base servers can be challenging. The tables indicate that Base64 encoding should be avoided and that there are no significant performance differences between using gRPC and REST.
+
+Two factors contribute to the similar performance results for gRPC and REST. Firstly, the model's inference time is considerably longer than the networking time, making it difficult to discern the small gains obtained by changing the transmission protocol. For example, sending 20 large images concurrently in the simple base case (Table 3) took roughly 0.19s, whereas we are now spending approximately 1.4 seconds, highlighting the significant impact of model inference time on the comparison.
+
+Secondly, the Torchserve implementation plays a role in these results. It has been observed that Torchserve's `.proto` definition for [prediction response](https://github.com/pytorch/serve/blob/master/frontend/server/src/main/resources/proto/inference.proto#L20-L23){:target="_blank"}{:rel="noopener noreferrer"} is too generic and it cannot be personalized with your model specifics.
+
+```python
+message PredictionResponse {
+    // Response content for prediction
+    bytes prediction = 1;
+}
+``` 
+
+This means that your response will be converted to a chunk of bytes so you would not be getting any advantage of Protobuf serialization (similar to what happens with images). A better or more customizable definition, as the one provided by Tfserving, of the `.proto` file could help to boost performance.
+
+## Conclusion
+
+If you're using Torchserve to serve computer vision models, it's recommended to steer clear of gRPC. Our findings show that there are no performance benefits to using gRPC, and it adds complexity to your code while making debugging more challenging due to its non-human-readable messages. Since REST is more commonly used, most developers are already familiar with it. Switching to gRPC in this scenario comes with a learning curve that doesn't offer any significant advantages.
```

Comment:
...benefits FROM using...

Reply:
Yo partiría la frase, es muy larga:

Our findings show that there are no performance benefits to using gRPC. Moreover, it adds code complexity while hindering debugging due to its non-human-readable messages